cpp/include/raft/neighbors/detail/cagra/add_nodes.cuh

```diff
@@ -0,0 +1,285 @@
+    handle, mr, raft::make_extents<std::int64_t>(max_chunk_size * base_degree, base_degree));
+
+  auto host_two_hop_neighbors_indices =
+    raft::make_host_matrix<IdxT, std::int64_t>(max_chunk_size * base_degree, base_degree);
+
+  auto two_hop_neighbors_distances = raft::make_device_mdarray<DistanceT, std::int64_t>(
+    handle, mr, raft::make_extents<std::int64_t>(max_chunk_size * base_degree, base_degree));
+
+  auto host_two_hop_neighbors_distances =
+    raft::make_host_matrix<DistanceT, std::int64_t>(max_chunk_size * base_degree, base_degree);
+
+  raft::spatial::knn::detail::utils::batch_load_iterator<T> additional_dataset_batch(
+    additional_dataset_view.data_handle(),
+    num_add,
+    additional_dataset_view.stride(0),
+    max_chunk_size,
+    raft::resource::get_cuda_stream(handle),
+    resource::get_workspace_resource(handle));
+  for (const auto& batch : additional_dataset_batch) {
+    // Step 1: Obtain K (=base_degree) nearest neighbors of the new vectors by CAGRA search
+    // Create queries
+    RAFT_CUDA_TRY(cudaMemcpy2DAsync(queries.data_handle(),
+                                    sizeof(T) * dim,
+                                    batch.data(),
+                                    sizeof(T) * additional_dataset_view.stride(0),
+                                    sizeof(T) * dim,
+                                    batch.size(),
+                                    cudaMemcpyDefault,
+                                    raft::resource::get_cuda_stream(handle)));
+
+    const auto queries_view = raft::make_device_matrix_view<const T, std::int64_t>(
+      queries.data_handle(), batch.size(), dim);
+
+    auto neighbor_indices_view = raft::make_device_matrix_view<IdxT, std::int64_t>(
+      neighbor_indices.data_handle(), batch.size(), base_degree);
+    auto neighbor_distances_view = raft::make_device_matrix_view<float, std::int64_t>(
+      neighbor_distances.data_handle(), batch.size(), base_degree);
+
+    raft::neighbors::cagra::search(
+      handle, params, idx, queries_view, neighbor_indices_view, neighbor_distances_view);
+
+    raft::copy(host_neighbor_indices.data_handle(),
+               neighbor_indices.data_handle(),
+               batch.size() * base_degree,
+               raft::resource::get_cuda_stream(handle));
+
+    // Step 2: Obtain K (=base_degree) nearest neighbors of the neighbors of the new vectors by
+    // CAGRA search
+    for (std::uint32_t i = 0; i < batch.size(); i++) {
+      for (std::uint32_t j = 0; j < base_degree; j++) {
+        raft::copy(
+          neighbors_vectors.data_handle() + (i * base_degree + j) * dim,
+          idx.dataset().data_handle() +
+            host_neighbor_indices.data_handle()[i * base_degree + j] * idx.dataset().stride(0),
+          dim,
+          raft::resource::get_cuda_stream(handle));
+      }
+    }
+
+    raft::neighbors::cagra::search(
+      handle, params, idx, queries_view, neighbor_indices_view, neighbor_distances_view);
```

Comment:
This uses the same `queries` as the search call earlier. Shouldn't it use `neighbor_vectors` as queries as input and `two_hop_neighbors_indices` as output instead?
```suggestion
      handle, params, idx, make_const_mdspan(neighbor_vectors.view()) two_hop_neighbors_indices.view(), two_hop_neighbors_distances.view();
```